### PR TITLE
YEARN07 Update - Instant Withdrawal 

### PR DESCRIPTION
**Pull request by 0xtaf**
# Instant Withdrawal 

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] New Bot
- [ ] Fix a bot error
- [x] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [ ] `npm install` command works (i.e. package-lock is updated properly)
- [ ] The `name` and `description` fields of `package.json` describe the bot properly
- [ ] All the libraries listed in `package.json` are being used
- [ ] All the development libraries are in `devDependencies` section of `package.json`
- [ ] The bot README is updated
- [ ] The bot README sufficiently describes what the bot does
- [ ] All the alerts that the bot emits are listed in the README
- [ ] The txn/block examples listed in the README generate the expected findings
- [ ] Code is easy to follow and/or properly commented
- [ ] Bot contains tests
- [ ] All tests pass when running `npm test`
- [ ] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [ ] Tests cover all the paths (i.e. negative & positive test cases)
- [ ] When the tests complete, no warning/errors are shown
- [ ] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [ ] The networks specified in the README and the networks are listed in the `package.json` are the same
- [ ] The code does what the README describes
- [ ] Network calls are properly being cached (if duplicated calls can occur)
- [ ] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [ ] The findings fields are filled out correctly
- [ ] Findings `protocol` is specified (if the bot is for a specific one)
- [ ] `metadata` of the findings contains all the information needed.
- [ ] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [ ] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [ ] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [ ] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [ ] `repository` is added in the `package.json`

## Comments (optional)

I removed ganache-core as it is [deprecated](https://www.npmjs.com/package/ganache-core) and used [ganache](https://github.com/trufflesuite/ganache) instead. To be able to use new ganache, I had to update `ganache.provider` options. `jsonRpcURL` field from the old `ganache-core` would throw error, so I removed `jsonRpcURL` variables from the code and used `network` instead by checking their [docs](https://github.com/trufflesuite/ganache#documentation).

### CODE REVIEW

File: Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts [hunk 1]
```diff
@@ -99,97 +96,129 @@ describe("DataFetcher tests suite", () => {
...
-          {inputs:[investor], outputs:[newAmount]},
-        )
+        .addCallFrom(vault, investor, "latest", vaultInterface, "balanceOf", {
+          inputs: [investor],
+          outputs: [newAmount],
+        })
+        .addCallFrom(vault, investor, "latest", vaultInterface, "balanceOf", {
+          inputs: [investor],
+          outputs: [newAmount],
+        })
```
[the remark this answers is not in the document]
Comment (RCantu92) [reply Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts:108]: On L196 and L200, they are different, which leads me to believe this unit test meant to have a similar set up.

File: Yearn-agents/InstantWithdrawal/src/agent.spec.ts [hunk 2]
```diff
@@ -2,32 +2,29 @@ import { Finding, FindingSeverity, FindingType, HandleBlock } from "forta-agent"
...
-    return newFindings;
-  };
+const updater = (findings: Finding[][]) => (n: number) => {
+  const newFindings: Finding[] = [];
+  for (let i = 0; i < n; ++i) newFindings.push(createGenericFinding(i));
+  findings.push(newFindings);
+  return newFindings;
+};
 
 describe("InstantWhitdraw agent tests suite", () => {
```
Comment (RCantu92) [inline Yearn-agents/InstantWithdrawal/src/agent.spec.ts:22]:
```suggestion
describe("Instant Withdraw agent tests suite", () => {
```
Small fix

**Review by RCantu92 — COMMENTED** [review]
Mostly small suggestions/fixes, and left some points that need some clarification.

Also, Is the the `url` in `makeRequestOptions` in `utils.ts` still valid? Just confirming in case we need to use another subgraph. (left the comment here since GH wouldn't let me a comment inline) We should look into this further.

File: Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts [hunk 3]
```diff
@@ -99,97 +96,129 @@ describe("DataFetcher tests suite", () => {
...
         .addSigner(investor)
         .getSigner(investor)
         .allowTransaction(
```
[the remark this answers is not in the document]
Comment (Vxatz) [reply Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts:111]: This test actually passes with only one `.addCallFrom` and the `.addSigner`.

File: Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts [hunk 4]
```diff
@@ -99,97 +96,129 @@ describe("DataFetcher tests suite", () => {
...
+        createAddress("0xda05"),
+        ["50", "100"],
+        ["50", "100"],
+      ],
     ];
 
-    for(let [numOfInvestors, jsonRpcURL, block, supply, vault, iBalance, fBalance] of CASES) {
+    for (let [
+      numOfInvestors,
+      jsonRpcURL,
```
Comment (Vxatz) [inline Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts:177]:
```suggestion
    
```

Comment (Vxatz) [reply Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts:177]: + One at L232

File: Yearn-agents/InstantWithdrawal/README.md [hunk 5]
```diff
@@ -2,22 +2,22 @@
...
 - YEARN-7
   - Fired at least every 6 hours.
   - Severity is always set to "Info".
   - Type is always set to "Info".
   - The metadata contains:
-    - `vault`: The vault analyzed. 
+    - `vault`: The vault analyzed.
     - `totalSupply`: Total supply of vault shares.
     - `sharesOwnedByTop10`: Amount of shares owned by the Top 10 investors.
     - `shareAvailableToWithdrawByTop10`: Amount of shares the the Top 10 investors was able to withdrawn instantly.
```
Comment (Vxatz) [inline Yearn-agents/InstantWithdrawal/README.md:21]:
```suggestion
    - `shareAvailableToWithdrawByTop10`: Amount of shares the Top 10 investors were able to withdraw instantly.
```

File: Yearn-agents/InstantWithdrawal/package.json [hunk 6]
```diff
@@ -1,7 +1,11 @@
 {
   "name": "yearn-7",
   "version": "0.0.1",
-  "description": "This agent informs about the amount of assets that can be instantanly withdrawn from vaults",
+  "description": "This bot informs about the amount of assets that can be instantanly withdrawn from vaults",
```
Comment (Vxatz) [inline Yearn-agents/InstantWithdrawal/package.json]:
```suggestion
  "description": "This bot informs about the amount of assets that can be instantly withdrawn from vaults",
```

File: Yearn-agents/InstantWithdrawal/src/agent.spec.ts (hunk 2 above)
Comment (Vxatz) [inline Yearn-agents/InstantWithdrawal/src/agent.spec.ts]:
```suggestion
describe("Instant Withdraw bot tests suite", () => {
```

File: Yearn-agents/InstantWithdrawal/src/data.fetcher.ts [hunk 7]
```diff
@@ -57,42 +53,41 @@ export default class DataFetcher {
...
+    const balancesPromises = investors.map(investor =>
       vaultContract.balanceOf(investor, { blockTag: "latest" })
     );
     const balances = await Promise.all(balancesPromises);
-  
+
     const totalInvestorValues = balances.reduce((acum, value) => acum.add(value));
-  
+
     let ableToWithdrawn = BigNumber.from(0);
```
Comment (Vxatz) [inline Yearn-agents/InstantWithdrawal/src/data.fetcher.ts:91]:
```suggestion
    let ableToWithdraw = BigNumber.from(0);
```

File: Yearn-agents/InstantWithdrawal/src/data.fetcher.ts [hunk 8]
```diff
@@ -102,31 +97,26 @@ export default class DataFetcher {
...
-    createFork: ForkFactory = getForkProvider,
+    createFork: ForkFactory = getForkProvider
   ): Promise<[string, BigNumber, BigNumber, BigNumber][]> {
     const blockNumber = await this.getBlockNumber();
     const vaults = await this.getVaults(registryAddress, blockNumber);
 
-    const investByVaultPromises = vaults.map((vault) =>
-      this.getBiggerInvestors(vault)
-    );
+    const investByVaultPromises = vaults.map(vault => this.getBiggerInvestors(vault));
```
Comment (Vxatz) [inline Yearn-agents/InstantWithdrawal/src/data.fetcher.ts:111]:
```suggestion
    const investorsByVaultPromises = vaults.map(vault => this.getBiggerInvestors(vault));
```

File: Yearn-agents/InstantWithdrawal/src/data.fetcher.ts [hunk 9]
```diff
@@ -102,31 +97,26 @@ export default class DataFetcher {
...
     let investByVault = await Promise.all(investByVaultPromises);
 
-    investByVault = investByVault.filter(
-      (investorsInfo) => investorsInfo.length > 0
-    );
+    investByVault = investByVault.filter(investorsInfo => investorsInfo.length > 0);
```
Comment (Vxatz) [inline Yearn-agents/InstantWithdrawal/src/data.fetcher.ts:114]:
```suggestion
    investorsByVault = investByVault.filter(investorsInfo => investorsInfo.length > 0);
```

**Review by Vxatz — COMMENTED** [review]
Non-logic fixes recommendations

File: Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xtaf) [reply Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts:108]: yes I think it's just duplicated. for L196 and 200, different balances are given and defined above them.

File: Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xtaf) [reply Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts:111]: @RCantu92 `.getSigner` can be called from [`MockEthersProvider`](https://github.com/NethermindEth/general-agents-module/blob/aa8e67bd15f80ab3f60ca33f25d957e7420ee930/src/mock.utils.ts#L13) and after getting the signer, `. allowTransaction` becomes callable.

@Vxatz that's right. one `.addCallFrom` passes the test.

File: Yearn-agents/InstantWithdrawal/package.json [hunk 10]
```diff
@@ -20,17 +24,16 @@
   },
   "dependencies": {
     "axios": "^0.21.4",
-    "forta-agent": "^0.0.39",
-    "ganache-core": "^2.13.2"
+    "forta-agent": "^0.1.9",
+    "ganache": "^7.4.3"
   },
   "devDependencies": {
```
Comment (RCantu92) [inline Yearn-agents/InstantWithdrawal/package.json:30]: It appears you have run `prettier`, but I don't see it installed under `devDependencies`.

File: Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (RCantu92) [reply Yearn-agents/InstantWithdrawal/src/data.fetcher.spec.ts:107]: Okay, I see. You can do that after calling `.addSigner()` in `MockEthersProvider` as shown [here](https://github.com/NethermindEth/general-agents-module/blob/main/src/test/mock_ethers_provider.ts#L100).

Great to know 👍

**Review by RCantu92 — APPROVED** [review]
Approved 👍 

Just left some minor comments.

File: Yearn-agents/InstantWithdrawal/package.json (hunk 10 above)
Comment (0xtaf) [reply Yearn-agents/InstantWithdrawal/package.json:30]: that's right. because I have it installed globally. so I forgot to add it here. added 👍

**Review by Vxatz — APPROVED** [review]

**Pull request merged**